### ISSUE TITLE
Permissions registry-type object for plugins?

### ISSUE THREAD

**Issue by Xemiru** [SpongePowered/SpongeAPI#601]
A way to get all the permissions of a plugin would be nice.

* A `PermissionData` class to store all information about a permission node's purpose. Should include information about the permission, the name (e.g. "/call"), the description of what it allows (not just "Allows access to the /call command."), a default setting, etc.

* A `Permissible` interface that can give the list of registered permissions of the plugin?
* A `PermissionRegistry` class holding `PermissionData` for plugins?
* Letting the `PluginContainer` hold all `PermissionData` for its plugin?
* boformer: Allow it as a parameter in the @Plugin annotation?

Use Cases:
* https://github.com/Xemiru/Permissions-Writer-Utility/issues/3
* Permissions plugins' tab completion

Pinging @zml2008 as he mentioned being able to do something like this.

**Comment by boformer** [comment]
It is part of the metadata of a plugin, so it should be a part of the @Plugin annotation.

I would prefer a simple builder for the permission data.

This has a really low priority for me.

**Comment by Xemiru** [comment]
Very low priority, but a feature nonetheless.

**Comment by MattBDev** [comment]
I really hope this can be more of a priority because for a plugins that involve protected blocks or land, they need to be able to register a permissions from outside of the command service. I see this as being critical for plugins like LWC, GriefPrevention, and PlotMe which all have the option to allow admins to override protections.

**Comment by zml2008** [comment]
uh permissions aren't registered at all currently why are you still talking about registering permissions?

**Comment by MattBDev** [comment]
It is something that is needed so developers of plugins similar to the ones I mentioned above can start working on their own plugins.

**Comment by zml2008** [comment]
Why do permissions need to be registered? There's nothing in sponge that requires registration of permissions?

**Comment by MattBDev** [comment]
How else are they supposed to be created? Since there is no plugin metadata file, there is no place to say what permissions a plugin has.

**Comment by zml2008** [comment]
permissions are a string. you pass the string to the hasPermission method. you don't need a plugin.yml to get strings?

**Comment by MattBDev** [comment]
So essentially any permission plugin will consider every string a permission regardless if there isn't a plugin on the server that has that permission?

**Comment by zml2008** [comment]
Well if a plugin checks for a permission then ofc it is a permission. simple as that

**Comment by boformer** [comment]
Yep. You can do hasPermission(randomString) and it would return true if the player has it. The same worked in Bukkit as well ;)

**Comment by ST-DDT** [comment]
This is issue is not about registering permissions.
Its just for providing a list of existing permissions with a little bit description so permission plugins can export them for better overview.
Basically some permission meta data.

**Comment by Xemiru** [comment]
Permission plugins at their core are only there to check whether or not a player has a given permission. It will not necessarily care about the extra information given to a permission, it will only give you a true or false value based on whether or not the target has the given permission node when applied to the rules of the plugin (default values, inheritance, etc).

This issue, as ST-DDT said, is only about helping provide access to extra information about a plugin's list of permissions should the need arise. In my case, it is to help collect all permissions available within a server's list of plugins in order to ease the task of finding these nodes. In a different plugin's case, say maybe a permissions plugin that also provides extra help, would have something like a `/perm info <node>` command to attain full information as to what the permission node itself gives to its holder.

**Comment by ryantheleach** [comment]
*If* this is implemented, I propose that it should generate a default configuration file, that allows defining what the permissions are. this file should not be bundled by the plugin, but lets administrators edit the permissions as opposed to hard-coding them.

Having a registered permissions system has a fatal flaw however, dynamic permissions.

How are you supposed to register a permission that isn't known before the check? or would you like to see a modifyworld.blocktypewithfullserialization permission registered for each blocktype/state ?

The only way I can see that working is if the prefix is registered, or a more complex template system which starts to violate some of the simplicity the current system has for little return.

**Comment by ST-DDT** [comment]
This seems too complicated.
Bundling an optional list/file of permissions with defaults and maybe a description that could be read by the permission service is probably enough. Or just a method to do so.
The itemwithxmetadata permission is a special permission that cannot be described well and should probably split into smaller pieces. But nethertheless it still needs a real documentation on a web page. It will never fit in the documentation you should pass along to the Permission API.

Summary: Optional way to register the permission.

**Comment by zml2008** [comment]
yeah @ryantheleach that is way too complicated. Never gonna happen.

And permission registration is designed to be a hint towards what permissions might be available, not a complete listing. Permissions plugins are not allowed to change their permission checking behavior based on what permissions are registered, since the system is entirely optional.

**Comment by ST-DDT** [comment]
Are there any ideas which data should be available/necessary for the registration?

* PermissionId
* default value suggestion ? (really not sure about this)
  - If provided the permission plugin can still overwrite/ignore it
*  security level suggestion (`String`: `Guest`, `User`, `Staff`, `Admin`, `Console`)
  - Human readable, not meant to be processed by the permission plugin
* description (`Text`)
* help/advanced documentation URL
  - Could be integrated in the description
* "owning" plugin ?
  - Maybe this is implicit due to the PermissionId

**Comment by boformer** [comment]
I would say plugin object, permission string and text description.

**Comment by Xemiru** [comment]
>permission registration is designed to be a hint towards what permissions might be available, not a complete listing.

rip my use case
i'd assume this means i'd have to garner all permissions another way, unless there's a method planned to be able to grab all possible permissions from a plugin

**Comment by zml2008** [comment]
All possible permissions from a plugin is all possible strings. Plugins would be expected to put in their best effort to give a good listing of possible permissions, but are not expected to try to register every single possibility (since this may be unknown until runtime)

**Comment by ST-DDT** [comment]
I wrote an [example class](https://gist.github.com/ST-DDT/951d54b156a3bcdc580f) how such an permission description could look like.

I slightly altered my original idea to be easier to use.

> 
    PermissionId
    default value suggestion ? (really not sure about this)
        If provided the permission plugin can still overwrite/ignore it
    security level suggestion (String: Guest, User, Staff, Admin, Console)
        Human readable, not meant to be processed by the permission plugin
    description (Text)
    help/advanced documentation URL
        Could be integrated in the description
    "owning" plugin ?
        Maybe this is implicit due to the PermissionId

`default value suggestion` is pretty useless with the `security level suggestion` aka `suggested rank`
`help/advanced documentation URL` should be included in the `description` `Text` object.

**Example Registration:**
````java
PermissionService service = ...;
PermissionDescription.Builder builder = service.getDescriptionBuilder(plugin);
builder.id("myPlugin.command.give.item.<ItemType>")
    .description(Text.of("Allows the player to use the <ItemType> in the /give command.\n"
        + "Use * to allow all ItemTypes.", 
        ClickAction.openUrl("example.com/MyPlugin/Permissions")))
    .suggestedRank(RANK_STAFF)
    .buildAndRegister();
````

**Example Usage:**
````
/perm info myPlugin.command.give
Allows access to the give command.
Click here for a more detailed description {OpenURL: "example.com/MyPlugin/Permissions"}
Suggested rank: staff
Owning plugin: MyPlugin {SendCommand: "/sponge plugin MyPlugin"}
Parent permissions: 
* myPlugin {SendCommand: "/perm info myPlugin", Hover: "Access to all MyPlugin's permissions"}
Child permissions: 
* myPlugin.command.give.item.<ItemType> {SendCommand: "/perm info myPlugin.command.give.item.<ItemType>", Hover: "Allows the player to use the <ItemType> in the /give command..."}
* myPlugin.command.give.enchantment.<Enchantment> {SendCommand: "/perm info myPlugin.command.give.enchantment.<Enchantment>", Hover: ...}
...
````
Plain:
````
/perm info myPlugin.command.give
Allows access to the give command.
Click here for a more detailed description
Suggested rank: staff
Owning plugin: MyPlugin
Parent permissions: 
* myPlugin
Child permissions: 
* myPlugin.command.give.item.<ItemType>
* myPlugin.command.give.enchantment.<Enchantment>
...
````

Any objections? If not then i'll open a PR for this.

**Comment by ST-DDT** [comment]
> * A `Permissible` interface that can give the list of registered permissions of the plugin?

I guess this or something similar is quite useful. @zml2008 Sorry that i haven't seen/covered it before.
(ex: `PermissionService.getDescriptions(PluginContainer) : Collection<PermissionDescription>`)

After that this issue can probably be closed.

**Comment by gabizou** [comment]
Closing as it was resolved in #747.
